lesson11/main.go

```diff
@@ -0,0 +1,107 @@
+package main
+
+import (
+	"fmt"
+	"sync"
+	"time"
+)
+
+type Fetcher interface {
+	// Fetch returns the body of URL and
+	// a slice of URLs found on that page.
+	Fetch(url string) (body string, urls []string, err error)
+}
+
+type checkV map[string]bool
+
+var V checkV = make(map[string]bool)
+var mux sync.Mutex
+
+func check(url string) bool {
+	mux.Lock()
+	defer mux.Unlock()
+	_, ok := V[url]
+	if ok == false {
+		V[url] = true
+		return false
+	}
+	return true
+}
+
+// Crawl uses fetcher to recursively crawl
+// pages starting with url, to a maximum of depth.
+func Crawl(url string, depth int, fetcher Fetcher) {
+	// TODO: Fetch URLs in parallel.
+	// TODO: Don't fetch the same URL twice.
+	// This implementation doesn't do either:
+	if depth <= 0 {
+		return
+	}
+	if check(url) {
+		return
+	}
+	body, urls, err := fetcher.Fetch(url)
+	if err != nil {
+		fmt.Println(err)
+		return
+	}
+	fmt.Printf("found: %s %q\n", url, body)
+	for _, u := range urls {
+		go Crawl(u, depth-1, fetcher)
+	}
+	return
+}
+
+func main() {
+	Crawl("http://golang.org/", 4, fetcher)
+	time.Sleep(5 * time.Second)
```

Comment:
I think this is not correct waiting all goroutines finish their work. Think about waitGroups usage